pkg/operator/ceph/object/objectstore.go

```diff
@@ -56,129 +62,460 @@ type idType struct {
 		lastStore = true
 	}
 
 	if !spec.PreservePoolsOnDelete {
-		err = deletePools(context, spec, lastStore)
+		err = deletePools(objContext, spec, lastStore)
 		if err != nil {
 			return errors.Wrap(err, "failed to delete object store pools")
 		}
 	} else {
-		logger.Infof("PreservePoolsOnDelete is set in object store %s. Pools not deleted", context.Name)
+		logger.Infof("PreservePoolsOnDelete is set in object store %s. Pools not deleted", objContext.Name)
 	}
+
 	return nil
 }
 
-func setMultisite(context *Context, serviceIP string, spec cephv1.ObjectStoreSpec, realmName string, zoneGroupName string, zoneName string) error {
-	realmArg := fmt.Sprintf("--rgw-realm=%s", realmName)
-	zoneGroupArg := fmt.Sprintf("--rgw-zonegroup=%s", zoneGroupName)
-	zoneArg := fmt.Sprintf("--rgw-zone=%s", zoneName)
-	endpointArg := fmt.Sprintf("--endpoints=%s:%d", serviceIP, spec.Gateway.Port)
-	updatePeriod := false
-
-	if spec.IsMultisite() {
-		updatePeriod = true
-		_, err := RunAdminCommandNoRealm(context, "zonegroup", "modify", realmArg, zoneGroupArg, endpointArg)
+// This is used for quickly getting the name of the realm, zone group, and zone for an object-store to pass into a Context
+func getMultisiteForObjectStore(context *clusterd.Context, store *cephv1.CephObjectStore) (string, string, string, error) {
+	if store.Spec.IsMultisite() {
+		zone, err := context.RookClientset.CephV1().CephObjectZones(store.Namespace).Get(store.Spec.Zone.Name, metav1.GetOptions{})
 		if err != nil {
-			return errors.Wrapf(err, "failed to add object store %q in rgw zone group %q", context.Name, zoneGroupName)
+			return "", "", "", errors.Wrapf(err, "error finding zone for %q", store.Name)
 		}
-		_, err = RunAdminCommandNoRealm(context, "zone", "modify", realmArg, zoneGroupArg, zoneArg, endpointArg)
+
+		zonegroup, err := context.RookClientset.CephV1().CephObjectZoneGroups(store.Namespace).Get(zone.Spec.ZoneGroup, metav1.GetOptions{})
 		if err != nil {
-			return errors.Wrapf(err, "failed to add object store %q in rgw zone %q", context.Name, zoneName)
+			return "", "", "", errors.Wrapf(err, "error finding zone group for %q", store.Name)
 		}
-		logger.Infof("Added object store %q to realm %q, zonegroup %q, zone %q", context.Name, realmName, zoneGroupName, zoneName)
-	} else {
-		// create the realm if it doesn't exist yet
-		output, err := runAdminCommand(context, "realm", "get")
+
+		realm, err := context.RookClientset.CephV1().CephObjectRealms(store.Namespace).Get(zonegroup.Spec.Realm, metav1.GetOptions{})
 		if err != nil {
+			return "", "", "", errors.Wrapf(err, "error finding realm for %q", store.Name)
+		}
+
+		return realm.Name, zonegroup.Name, zone.Name, nil
+	}
+
+	return store.Name, store.Name, store.Name, nil
+}
+
+func checkZoneIsMaster(objContext *Context) (bool, error) {
+	logger.Debugf("checking if zone %v is the master zone", objContext.Zone)
+	realmArg := fmt.Sprintf("--rgw-realm=%s", objContext.Realm)
```

Comment:
Can we have a helper function to build the multiste args? It seems we are repeating this pattern.